File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Raytracer/Raytracer_VisualBasic/Plane.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: Plane.vb
'
'--------------------------------------------------------------------------

Public Class Plane
    Inherits SceneObject
    Public Norm As Vector
    Public Offset As Double

    Public Sub New(ByVal norm As Vector, ByVal offset As Double, ByVal surface As Surface)
        MyBase.New(surface)
        Me.Norm = norm
        Me.Offset = offset
    End Sub

    Public Overrides Function Intersect(ByVal ray As Ray) As ISect
        Dim denom = Vector.Dot(Norm, ray.Direction)
        If denom > 0 Then Return Nothing
        Return New ISect(Me, ray, (Vector.Dot(Norm, ray.Start) + Offset) / (-denom))
    End Function

    Public Overrides Function Normal(ByVal position As Vector) As Vector
        Return Norm
    End Function

End Class

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Raytracer/Raytracer_CSharp/Plane.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: Plane.cs
//
//--------------------------------------------------------------------------

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples
{
    class Plane : SceneObject
    {
        public Vector Norm;
        public double Offset;

        public Plane(Vector norm, double offset, Surface surface) : base(surface) { Norm = norm; Offset = offset; }

        public override ISect Intersect(Ray ray)
        {
            double denom = Vector.Dot(Norm, ray.Dir);
            if (denom > 0) return ISect.Null;
            return new ISect(this, ray, (Vector.Dot(Norm, ray.Start) + Offset) / (-denom));
        }

        public override Vector Normal(Vector pos)
        {
            return Norm;
        }
    }
}
